proposals/0000-overloaded-bracket.rst

```diff
@@ -0,0 +1,239 @@
+
+   class Applicative m => Quote m where
+      newName :: String -> m Name
+      fail :: String -> m Name
+
+An instance for ``Quote`` can be implemented
+for ``Q``, retaining backwards compatibility, but also ``State NameSupply`` to
+enable a pure way to extract expressions.
+
+``Q`` is not necessary for the implementation of quotations.
+It provides much more power than is necessary. Quotations are used in order to
+construct a representation of expressions, the only effect which is used in the
+implementation is a name generation effect.
+
+Detaching quotations from ``Q`` makes way for a form of "pure" template haskell
+which can be fully evaluated on the host without having to run effects on the target.
+
+
+Proposed Change Specification
+-----------------------------
+
+The goal of the changes is for an expression ``e : T`` to give the representation
+``[| e |] : Quote m => m Exp``. Several steps are necessary to make this change possible.
+
+1. Define the interface for ``Quote``::
+
+      class Applicative m => Quote m where
+         newName :: String -> m Name
+         fail :: String -> m Name
+
+   These are all the operations which are necessary to build the representation
+   of expressions.
+
+2. Generalise all the combinators to build syntax in ``Language.Haskell.TH.Lib``.
+   Due to an `audit <https://github.com/ghc-proposals/ghc-proposals/issues/211#issuecomment-472092412>`_ conducted by Richard, it was found that the only effect from
+   ``Q`` which was used is the ``newName`` function which generates a fresh name.
+   All the other combinators can be defined using the ``Applicative`` operations.
+
+
+3. Generalise the ``Lift`` type class::
+
+      class Lift a where
+         lift :: Quote m => a -> m Exp
+         liftT :: Quote m => a -> m (TExp a)
+
+   This is necessary so that implicit lifting can continue to work without
+   enforcing strong constraints on the type of the bracket.
+
+4. Refine the rules to do with splicing. A top-level splice instantiates ``m ~ Q``
+   and operates as before. The use of nested brackets doesn't enforce any
+   specific constraints on ``m``. A quotation which contains splices inherits
+   the constraints on the representation.
+
+Proposed Change Specification (Typed Template Haskell)
+------------------------------------------------------
+
+There is another useful axis of polymorphism for typed quotations.
+It is useful to be polymorphic in the type constructor
+``p``. This can be used to embed quotations directly into an EDSL and write
+instances for typed quotations which isn't convenient without the polymorpism.
+
+Example 1: Creating a partially static data type::
+
+      data Expr m a where
+         Leaf :: Code m a -> Expr a
+         Int :: Int -> Expr Int
+         Add :: Expr Int -> Expr Int -> Expr Int
+
+      data Code m a = Code (m (TExp a))
+
+      instance TypedQuote Expr where
+         fromBracket = Leaf . Code
+
+      expr = [|| 5 ||] `Add` [|| 10 ||]
```

Comment:
Admittedly, I don't fully understand this use case. OK, now we have `expr :: Quote m => Expr m Int`, how would one instantiate `m` and make use of this `expr`?

And why would one prefer it over this variant, where `Expr` does not mention `m`:

```
data Expr a where
  Leaf :: TExp a -> Expr a
  Int :: Int -> Expr Int
  Add :: Expr Int -> Expr Int -> Expr Int

newtype ExprGen m a =
  ExprGen (m (Expr a))  -- like Code for TExp

instance TypedQuote ExprGen where
  fromBracket = ExprGen . fmap Leaf

add :: Applicative m => ExprGen m Int -> ExprGen m Int -> ExprGen m Int
add (ExprGen x) (ExprGen y) = ExprGen (liftA2 Add x y)

expr = [|| 5 ||] `add` [|| 10 ||]
```

Reply:
That code also seems fine. I think you could choose to design it either way?

Reply:
Ah, but then there's `fmap`, so this class

```
class TypedQuote p where
   fromBracket :: m (TExp a) -> p m a
```

must become

```
class TypedQuote p where
   fromBracket :: Functor m => m (TExp a) -> p m a
```